service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/CreateAzureVmStep.java

```diff
@@ -261,9 +263,27 @@ private NetworkInterface createNetworkInterface(
             .withExistingPrimaryNetwork(existingNetwork)
             .withSubnet(subnetName)
             .withPrimaryPrivateIPAddressDynamic();
-    if (existingAzureIp.isPresent()) {
-      createNicStep.withExistingPrimaryPublicIPAddress(existingAzureIp.get());
-    }
+    existingAzureIp.ifPresent(createNicStep::withExistingPrimaryPublicIPAddress);
     return createNicStep.create();
   }
+
+  private NetworkInterface createOrGetExistingNetworkInterface(
+      FlightContext context,
+      ComputeManager computeManager,
+      AzureCloudContext azureCloudContext,
+      Network existingNetwork,
+      String subnetName,
+      Optional<PublicIpAddress> existingAzureIp) {
+    NetworkInterface networkInterface;
+    if (context.getWorkingMap().containsKey(WORKING_MAP_NETWORK_INTERFACE_KEY)) {
+      String nicId = context.getWorkingMap().get(WORKING_MAP_NETWORK_INTERFACE_KEY, String.class);
+      networkInterface = computeManager.networkManager().networkInterfaces().getById(nicId);
+    } else {
+      networkInterface =
+          createNetworkInterface(
+              computeManager, azureCloudContext, existingNetwork, subnetName, existingAzureIp);
+      context.getWorkingMap().put(WORKING_MAP_NETWORK_INTERFACE_KEY, networkInterface.id());
```

Comment:
I do not recommend this technique, although it probably works most of the time.

The workingMap is only serialized at a step boundary, so it is not persisted here on retry. You would do just as well with a local variable. You can see, though, that on a pod failure or failover, the working map would be state is lost.

I think a better approach would be to make a new step that just handles the create of the network interface. When that step completes, you would know you had created it and that the working map item would be persisted.

Reply:
Thank you for the feedback. I probably missed that part. Let me check it.

Reply:
New step for network interface added